panic on switching from large file to next file that is smaller

I am viewing multiple files. One is large, and I scroll to the end. Then I type :n to get to next file which is shorter, and I get a panic. I don't get a panic when I scroll back to the top before switching to the next file.

moor version v2.13.3

```
Panic recovery timestamp: 2026-05-30 23:22:49.401347028 +0200 CEST m=+15.791056049

panic: scrollPosition index={70} not found in allLines size 48, indexed {70}-{117}:
        :
        (internal.scrollPosition) {
         internalDontTouch: (internal.scrollPositionInternal) {
          lineIndex: (*linemetadata.Index)(0x2d7430e24830)({
           index: (int) 60
          }),
          delta: (linemetadata.ScreenLines) 0,
          name: (string) (len=14) "Pager Makefile",
          canonicalizing: (bool) false,
          canonical: (internal.scrollPositionCanonical) {
           width: (int) 190,
           height: (linemetadata.ScreenLines) 48,
           showLineNumbers: (bool) true,
           showStatusBar: (bool) true,
           wrapLongLines: (bool) false,
           pagerLineCount: (int) 108,
           lineIndex: (*linemetadata.Index)(0x2d7430e24830)(<already shown>),
           deltaScreenLines: (linemetadata.ScreenLines) 0
          }
         }
        }
         [recovered, repanicked]

goroutine 1 [running]:
main.main.func1()
        github.com/walles/moor/v2/cmd/moor/moor.go:691 +0x1d4
panic({0x6b2880?, 0x2d743127c2d0?})
        runtime/panic.go:860 +0x13a
main.startPaging.func1()
        github.com/walles/moor/v2/cmd/moor/moor.go:754 +0xb8
panic({0x6b2880?, 0x2d743127c2d0?})
        runtime/panic.go:860 +0x13a
github.com/walles/moor/v2/internal.(*Pager).internalRenderLines(0x2d7430f2a248, 0x1)
        github.com/walles/moor/v2/internal/screenLines.go:171 +0xadc
github.com/walles/moor/v2/internal.(*Pager).renderLines(0x2d7430f2a248)
        github.com/walles/moor/v2/internal/screenLines.go:88 +0x54
github.com/walles/moor/v2/internal.(*Pager).redraw(0x2d7430f2a248, {0x0, 0x0})
        github.com/walles/moor/v2/internal/screenLines.go:50 +0xa8
github.com/walles/moor/v2/internal.(*Pager).StartPaging(0x2d7430f2a248, {0x956c10, 0x2d7430f29560}, 0x2d7430f79ec0, 0x2d74311e1de0)
        github.com/walles/moor/v2/internal/pager.go:638 +0x32c
main.startPaging(0x0?, {0x956c10?, 0x2d7430f29560?}, 0x9?, 0x9?)
        github.com/walles/moor/v2/cmd/moor/moor.go:762 +0x53
main.main()
        github.com/walles/moor/v2/cmd/moor/moor.go:718 +0x267
```

```
Version      : v2.13.3                                                                                                                                                                        
LANG         : nl_NL.UTF-8                                                                                                                                                                    
TERM         : tmux-256color                                                                                                                                                                  
EDITOR       : nvim                                                                                                                                                                           
TERM_PROGRAM : tmux                                                                                                                                                                           
                                                                                                                                                                                              
GOOS    : linux                                                                                                                                                                               
GOARCH  : amd64                                                                                                                                                                               
Compiler: gc                                                                                                                                                                                  
NumCPU  : 16                                                                                                                                                                                  
                                                                                                                                                                                              
Stdin  is a terminal: true                                                                                                                                                                    
Stdout is a terminal: true                                                                                                                                                                    
                                                                                                                                                                                              
MOOR: -style lovelace -wrap -no-linenumbers -reformat                                                                                                                                         
Commandline: []string{"moor", "-trace", "mdview.c", "mdview.go"}                                                                                                                              
                                                                                                                                                                                              
time="May 31 13:02:17.466294" level=debug msg="File is assumed to be uncompressed: mdview.c"                                                                                                  
time="May 31 13:02:17.468601" level=debug msg="File is assumed to be uncompressed: mdview.go"                                                                                                 
time="May 31 13:02:17.468730" level=debug msg="File is assumed to be uncompressed: mdview.c"                                                                                                  
time="May 31 13:02:17.468758" level=debug msg="Counted 118 lines in 10.961µs at 92ns/line"                                                                                                    
time="May 31 13:02:17.468774" level=info msg="Stream read in 85.337µs, have 118 lines"                                                                                                        
time="May 31 13:02:17.471206" level=debug msg="File is assumed to be uncompressed: mdview.go"                                                                                                 
time="May 31 13:02:17.471243" level=info msg="ttyin terminal state: &{state:{termios:{Iflag:16384 Oflag:4 Cflag:191 Lflag:2608 Line:0 Cc:[3 28 127 21 4 0 1 0 17 19 26 0 18 15 23 22 0 0 0] Is
peed:0 Ospeed:0}}}"                                                                                                                                                                           
time="May 31 13:02:17.471246" level=debug msg="Counted 108 lines in 25.051µs at 231ns/line"                                                                                                   
time="May 31 13:02:17.471253" level=info msg="ttyout terminal state: &{state:{termios:{Iflag:16384 Oflag:4 Cflag:191 Lflag:2608 Line:0 Cc:[3 28 127 21 4 0 1 0 17 19 26 0 18 15 23 22 0 0 0] I
speed:0 Ospeed:0}}}"                                                                                                                                                                          
time="May 31 13:02:17.471264" level=info msg="GNOME Terminal detected, assuming arrow keys emulation active"                                                                                  
time="May 31 13:02:17.471271" level=info msg="Stream read in 82.603µs, have 108 lines"                                                                                                        
time="May 31 13:02:17.471272" level=debug msg="Using style <lovelace>"                                                                                                                        
time="May 31 13:02:17.471291" level=info msg="Entering Twin main loop..."                                                                                                                     
time="May 31 13:02:17.471385" level=debug msg="102 lines of search history processed from /home/peter/.lesshst"                                                                               
time="May 31 13:02:17.471401" level=debug msg="Removed 9 redundant search history lines"                                                                                                      
time="May 31 13:02:17.471404" level=info msg="Imported 91 search history entries from less"                                                                                                   
time="May 31 13:02:17.471423" level=info msg="Pager starting"                                                                                                                                 
time="May 31 13:02:17.471503" level=debug msg="Terminal background color detected as #c0d0d0 after 221.201µs"                                                                                 
time="May 31 13:02:17.475560" level=trace msg="Reader done, contents explicitly set"                                                                                                          
time="May 31 13:02:17.475566" level=debug msg="highlightFromMemory() took 4.275232ms"                                                                                                         
time="May 31 13:02:17.475577" level=debug msg="Tailing file mdview.go"
time="May 31 13:02:17.476609" level=trace msg="Line numbers style set from Chroma: #7f7f7f on #c0d0d0"
time="May 31 13:02:17.476616" level=trace msg="Plain text style set from Chroma: Default color on #c0d0d0"
time="May 31 13:02:17.476619" level=trace msg="Search hit style set to default: reverse Default color on Default color"
time="May 31 13:02:17.476622" level=trace msg="Search hit line background set to mixed color: #9aa6a6"
time="May 31 13:02:17.476625" level=trace msg="Pager: Setting target line to <nil>..."
time="May 31 13:02:17.476626" level=trace msg="Setting pause-after-lines to 50000..."
time="May 31 13:02:17.476631" level=info msg="Entering pager main loop..."
time="May 31 13:02:17.476632" level=trace msg="redraw called"
time="May 31 13:02:17.476635" level=info msg="Screen height changed from 0 to 49"
time="May 31 13:02:17.477572" level=trace msg="Reader done, contents explicitly set"
time="May 31 13:02:17.477577" level=debug msg="highlightFromMemory() took 8.798949ms"
time="May 31 13:02:17.477591" level=debug msg="Tailing file mdview.c"
time="May 31 13:02:17.478634" level=trace msg="redraw called"
time="May 31 13:02:17.677592" level=trace msg="redraw called"
time="May 31 13:02:18.476641" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:18.476711" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:18.477836" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:18.477886" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:18.873033" level=debug msg="ttyin high watermark bumped to 4 bytes"
time="May 31 13:02:18.873176" level=trace msg="Handling key event 15..."
time="May 31 13:02:18.878249" level=trace msg="Pager: Setting target line to <nil>..."                                                                                                        
time="May 31 13:02:18.878255" level=trace msg="Setting pause-after-lines to 50000..."
time="May 31 13:02:18.878257" level=trace msg="redraw called"
time="May 31 13:02:19.097367" level=trace msg="Handling key event 15..."
time="May 31 13:02:19.101392" level=trace msg="Pager: Setting target line to &{9223372036854775807}..."
time="May 31 13:02:19.101398" level=trace msg="Setting pause-after-lines to 9223372036854775807..."
time="May 31 13:02:19.101400" level=trace msg="redraw called"
time="May 31 13:02:19.273481" level=trace msg="Handling key event 15..."
time="May 31 13:02:19.278789" level=trace msg="Pager: Setting target line to &{9223372036854775807}..."
time="May 31 13:02:19.278794" level=trace msg="Setting pause-after-lines to 9223372036854775807..."
time="May 31 13:02:19.278796" level=trace msg="redraw called"
time="May 31 13:02:19.473165" level=trace msg="Handling key event 15..."
time="May 31 13:02:19.476872" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:19.476929" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:19.478076" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:19.478088" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:19.478164" level=trace msg="Pager: Setting target line to &{9223372036854775807}..."
time="May 31 13:02:19.478170" level=trace msg="Setting pause-after-lines to 9223372036854775807..."
time="May 31 13:02:19.478173" level=trace msg="redraw called"
time="May 31 13:02:20.278540" level=trace msg="Handling rune event '\"'/0x0022..."
time="May 31 13:02:20.278554" level=debug msg="Unhandled rune keypress '\"'/0x00000022"
time="May 31 13:02:20.278556" level=trace msg="redraw called"
time="May 31 13:02:20.477655" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:20.477679" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:20.478701" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:20.478714" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:21.134651" level=trace msg="Handling rune event 'n'/0x006e..."
time="May 31 13:02:21.134661" level=trace msg="redraw called"
time="May 31 13:02:21.478511" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:21.478579" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:21.479636" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:21.479684" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:22.479190" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:22.479245" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:22.480324" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:22.480343" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:23.480246" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:23.480271" level=trace msg="File mdview.go unchanged at 1767 bytes, continue tailing"
time="May 31 13:02:23.481401" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:23.481420" level=trace msg="File mdview.c unchanged at 4068 bytes, continue tailing"
time="May 31 13:02:23.537243" level=trace msg="Handling rune event ':'/0x003a..."
time="May 31 13:02:23.537272" level=trace msg="Pager: Setting target line to <nil>..."
time="May 31 13:02:23.537282" level=trace msg="Setting pause-after-lines to 50000..."
time="May 31 13:02:23.537290" level=trace msg="redraw called"
time="May 31 13:02:24.321348" level=trace msg="Handling rune event 'n'/0x006e..."
time="May 31 13:02:24.321383" level=trace msg="Switched to next file, index 1"
time="May 31 13:02:24.321397" level=trace msg="redraw called"
time="May 31 13:02:24.325878" level=info msg="Panic detected, closing screen before informing the user..."
time="May 31 13:02:24.325899" level=info msg="Interruptable reader interrupted"

Panic recovery timestamp: 2026-05-31 13:02:24.326206602 +0200 CEST m=+6.869513849

panic: scrollPosition index={70} not found in allLines size 48, indexed {70}-{117}:
        :
        (internal.scrollPosition) {
         internalDontTouch: (internal.scrollPositionInternal) {
          lineIndex: (*linemetadata.Index)(0x2a5fef01b4a8)({
           index: (int) 60
          }),
          delta: (linemetadata.ScreenLines) 0,
          name: (string) (len=14) "Pager mdview.c",
          canonicalizing: (bool) false,
          canonical: (internal.scrollPositionCanonical) {
           width: (int) 190,
           height: (linemetadata.ScreenLines) 48,
           showLineNumbers: (bool) false,
           showStatusBar: (bool) true,
           wrapLongLines: (bool) true,
           pagerLineCount: (int) 108,
           lineIndex: (*linemetadata.Index)(0x2a5fef01b4a8)(<already shown>),
           deltaScreenLines: (linemetadata.ScreenLines) 0
          }
         }
        }
         [recovered, repanicked]

goroutine 1 [running]:
main.main.func1()
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/cmd/moor/moor.go:691 +0x1d4
panic({0x6b2880?, 0x2a5fef3ac030?})
        /my/opt/go/src/runtime/panic.go:860 +0x13a
main.startPaging.func1()
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/cmd/moor/moor.go:754 +0xb8
panic({0x6b2880?, 0x2a5fef3ac030?})
        /my/opt/go/src/runtime/panic.go:860 +0x13a
github.com/walles/moor/v2/internal.(*Pager).internalRenderLines(0x2a5fef000248, 0x1)
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/internal/screenLines.go:171 +0xadc
github.com/walles/moor/v2/internal.(*Pager).renderLines(0x2a5fef000248)
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/internal/screenLines.go:88 +0x54
github.com/walles/moor/v2/internal.(*Pager).redraw(0x2a5fef000248, {0x0, 0x0})
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/internal/screenLines.go:50 +0xa8
github.com/walles/moor/v2/internal.(*Pager).StartPaging(0x2a5fef000248, {0x956c10, 0x2a5feef7ca20}, 0x2a5fef2a8720, 0x2a5fef2b12d0)
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/internal/pager.go:638 +0x32c
main.startPaging(0x0?, {0x956c10?, 0x2a5feef7ca20?}, 0x4?, 0x4?)
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/cmd/moor/moor.go:762 +0x53
main.main()
        /home/peter/go/pkg/mod/github.com/walles/moor/v2@v2.13.3/cmd/moor/moor.go:718 +0x267
```

--- a/internal/file-switching.go
+++ b/internal/file-switching.go
@@ -12,7 +12,7 @@ func (p *Pager) previousFile() {
 	if newIndex < 0 {
 		newIndex = 0
 	}
-	p.currentReader = newIndex
+	p.switchToFile(newIndex)
 	log.Tracef("Switched to previous file, index %d", p.currentReader)
 
 	select {
@@ -29,7 +29,7 @@ func (p *Pager) nextFile() {
 	if newIndex >= len(p.readers) {
 		newIndex = len(p.readers) - 1
 	}
-	p.currentReader = newIndex
+	p.switchToFile(newIndex)
 	log.Tracef("Switched to next file, index %d", p.currentReader)
 
 	select {
@@ -42,11 +42,20 @@ func (p *Pager) firstFile() {
 	p.readerLock.Lock()
 	defer p.readerLock.Unlock()
 
-	p.currentReader = 0
+	p.switchToFile(0)
 	log.Tracef("Switched to first file, index %d", p.currentReader)
 
 	select {
 	case p.readerSwitched <- struct{}{}:
 	default:
 	}
+}
+
+func (p *Pager) switchToFile(newIndex int) {
+	if newIndex == p.currentReader {
+		return
+	}
+
+	p.currentReader = newIndex
+	p.scrollPosition = newScrollPosition("Pager file switch")
 }

--- a/internal/file-switching_test.go
+++ b/internal/file-switching_test.go
@@ -1,0 +1,45 @@
+package internal
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/walles/moor/v2/internal/linemetadata"
+	"github.com/walles/moor/v2/internal/reader"
+	"github.com/walles/moor/v2/twin"
+	"gotest.tools/v3/assert"
+)
+
+func TestIssue426SwitchingToSmallerFileClearsStaleScrollPosition(t *testing.T) {
+	largeReader := reader.NewFromTextForTesting("large", strings.Repeat("large\n", 118))
+	smallReader := reader.NewFromTextForTesting("small", strings.Repeat("small\n", 48))
+	assert.NilError(t, largeReader.Wait())
+	assert.NilError(t, smallReader.Wait())
+
+	pager := NewPager(largeReader, smallReader)
+	pager.screen = twin.NewFakeScreen(190, 49)
+	pager.WrapLongLines = true
+	pager.ShowLineNumbers = false
+	pager.showLineNumbers = false
+	pager.filteringReader.SetBackingReader(smallReader)
+
+	// This matches the stale post-switch state from issue #426: the reader has
+	// moved to a shorter file, but the scroll position still points below it.
+	staleIndex := linemetadata.IndexFromZeroBased(70)
+	staleIndexPtr := &staleIndex
+	pager.scrollPosition = scrollPosition{
+		internalDontTouch: scrollPositionInternal{
+			lineIndex: staleIndexPtr,
+			delta:     0,
+			name:      "Issue 426 stale file switch",
+		},
+	}
+	pager.scrollPosition.internalDontTouch.canonical = canonicalFromPager(pager)
+
+	pager.nextFile()
+
+	rendered := pager.renderLines()
+	assert.Assert(t, len(rendered.lines) > 0)
+	assert.Equal(t, renderedToString(rendered.lines[0].cells), "small")
+	assert.Equal(t, pager.lineIndex().Index(), 0)
+}

